### PR TITLE
Validate all data scenarios

### PR DESCRIPTION
**Pull request by pjlee11**
_Asynchronously reads the directory `/data` and then synchronously validates each JSON file it finds._

Still requires testing and validating the `data/prod/` dir

* [x] Fixes https://github.com/bbc/simorgh/issues/287
* [x] Tests added for new features

* ~[ ] Test engineer approval~

### CODE REVIEW

File: src/app/dataValidator/index.test.js [hunk 1]
```diff
@@ -1,13 +1,79 @@
...
+
+  it('should ignore scenario-23.json', () => {
+    const filenames = ['scenario-23.json'];
+
+    readFiles(filenames);
+
+    expect(readFileSyncMock).not.toBeCalled();
+  });
+
+  it('should ignore scenario-23.json', () => {
```
Comment (sareh) [inline src/app/dataValidator/index.test.js]: Test description should be clarified: `it('should ignore files that are not of type ".json"', `

File: src/app/dataValidator/index.js [hunk 2]
```diff
@@ -9,8 +10,43 @@ const validateData = data => {
...
 
-const data = require('../../../data/scenario-01.json');
+const validateFile = filename => {
+  const data = fs.readFileSync(filename);
+  validateData(JSON.parse(data));
+};
+
+const readAllFiles = (filenames, dirname) =>
+  filenames.forEach(filename => {
+    // explicitly ignore scenario-23
```
Comment (sareh) [inline src/app/dataValidator/index.js]: Would be good to say why this scenario is being ignored - since it's deliberately an example of invalid data.

File: src/app/dataValidator/index.js [hunk 3]
```diff
@@ -9,8 +10,43 @@ const validateData = data => {
...
+    if (filename.includes('scenario-23.json')) {
+      return;
+    }
+
+    if (filename.includes('.json')) {
```
Comment (sareh) [inline src/app/dataValidator/index.js]: It would be better to use `path.extname(filename) === '.json'` https://nodejs.org/api/path.html#path_path_extname_path

File: src/app/dataValidator/index.js [hunk 4]
```diff
@@ -9,8 +10,43 @@ const validateData = data => {
...
+      validateFile(`${dirname}/${filename}`);
+    }
+  });
```
Comment (sareh) [inline src/app/dataValidator/index.js]: It would be good to alert that there is an invalid file if a non-`.json` file is in the directory.

File: src/app/dataValidator/index.js [hunk 5]
```diff
@@ -9,8 +10,43 @@ const validateData = data => {
...
 
-validateData(data);
+const readdirAsync = dirname =>
+  new Promise((resolve, rejects) => {
+    fs.readdir(dirname, (err, filenames) => {
+      if (err) {
+        rejects(err);
+      } else {
+        resolve(readAllFiles(filenames, dirname));
```
Comment (sareh) [inline src/app/dataValidator/index.js]: I think that here this has a single promise that will return for the group of files. 
I think it would be better to use `Promise.all()` so that this Promise resolves when all of the Promises in the iterable resolve (i.e. all files are validated).

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

File: src/app/dataValidator/index.js [hunk 6]
```diff
@@ -9,8 +10,43 @@ const validateData = data => {
...
+      }
+    });
+  });
 
-module.exports.validateData = validateData;
+console.time('readdirAsync'); // eslint-disable-line no-console
```
Comment (sareh) [inline src/app/dataValidator/index.js]: Since this is just for feedback to the user, `console.time` is great. 
If we want to use timing to measure performance of the validation in future, we should consider using the Performance API with `performance.mark()` and `performance.measure()` instead.

File: src/app/dataValidator/index.js [hunk 7]
```diff
@@ -9,8 +10,43 @@ const validateData = data => {
...
+readdirAsync('./././data').then(() => {
+  console.timeEnd('readdirAsync'); // eslint-disable-line no-console
+  log('\nAll files validated!');
+});
+
+module.exports = {
```
Comment (sareh) [inline src/app/dataValidator/index.js]: I think in this file you should only export a single function. 
If you want to export each of these for testing purposes, then they should be in separate files.

File: src/app/dataValidator/index.js (hunk 5 above)
Comment (sareh) [reply src/app/dataValidator/index.js]: I saw that in your PR description that you want to synchronously validate the json files - why is that? It seems better to do it asynchronously.

File: src/app/dataValidator/index.test.js [hunk 8]
```diff
@@ -1,13 +1,94 @@
...
   it('should not error on validateData', () => {
     expect(() => {
       validateData(data);
     }).not.toThrowError();
   });
 });
+
+describe('Data Validator requires mocking', () => {
+  beforeEach(() => {
+    readFileSyncMock = jest.spyOn(fs, 'readFileSync');
```
Comment (sareh) [inline src/app/dataValidator/index.test.js]: Great use of spying and mocking in this file. 👍

**Review by sareh — CHANGES_REQUESTED** [review]

File: src/app/dataValidator/index.js [hunk 9]
```diff
@@ -9,8 +10,43 @@ const validateData = data => {
...
+        rejects(err);
+      } else {
+        resolve(readAllFiles(filenames, dirname));
+      }
+    });
+  });
 
-module.exports.validateData = validateData;
+console.time('readdirAsync'); // eslint-disable-line no-console
+readdirAsync('./././data').then(() => {
```
Comment (jtart) [inline src/app/dataValidator/index.js]: Probably out of scope of this PR, but the script is very loud now that we test all files, we should put verbose logging behind a debugging flag.

File: data/schema.yaml [hunk 10]
```diff
@@ -93,8 +93,14 @@ components:
             articleType:
               enum:
                 - analysis
+                - askTheAudience
```
Comment (jtart) [inline data/schema.yaml]: What is the context for adding these enums? Can they be added in the separate PR? We should keep adding to our schema distinct from other pieces of work.

File: src/app/dataValidator/index.js (hunk 4 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: The schema lives here and is a `.yaml` so I don't think we need to be explicit that only `.json` files are allowed in this directory

Comment (sareh) [reply src/app/dataValidator/index.js]: OK, then this method shouldn't be called `readAllFiles` but `readScenarios` or something along those lines.

File: src/app/dataValidator/index.js (hunk 7 above)
Comment (jtart) [reply src/app/dataValidator/index.js]: Yes, we could use something like [rewire](https://github.com/jhnns/rewire) if we need to test internals. Unsure if there is anything newer in the new Jest world (Rewire is testing lib agnostic though).

**Review by jtart — CHANGES_REQUESTED** [review]

File: src/app/dataValidator/index.js (hunk 9 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: This will be handled as part of https://github.com/bbc/simorgh/issues/239

File: data/schema.yaml (hunk 10 above)
Comment (pjlee11) [reply data/schema.yaml]: These types are part of the current scenarios but were never added to the schema, without them the validation fails and therefore the test suite can't run and Travis complains and I would have to `--no-verify` to push any changes. 
It is likely they will be updated within https://github.com/bbc/simorgh/issues/380 which will be merged before this PR.

File: src/app/dataValidator/index.js (hunk 7 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: The commit https://github.com/bbc/simorgh/pull/340/commits/1125839d944889df7279fa461f2063bb7eb7c44a splits these out into more granular methods and which makes the testing/exporting more explicit. 
One part to mention is where I prepend `module.exports.` to the method call (https://github.com/bbc/simorgh/pull/340/commits/1125839d944889df7279fa461f2063bb7eb7c44a#diff-82c901e3afd825f4a1151b3b0f510687R13) this is needed otherwise the spy can't track calls to a method in the current module.

File: src/app/dataValidator/index.js (hunk 6 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: Will look to implement this in a future PR

File: src/app/dataValidator/index.test.js (hunk 1 above)
Comment (pjlee11) [reply src/app/dataValidator/index.test.js]: 👍

File: src/app/dataValidator/index.js (hunk 2 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: 👍

File: src/app/dataValidator/index.js (hunk 3 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: 👍

File: src/app/dataValidator/index.js (hunk 4 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: 👍

File: src/app/dataValidator/index.js (hunk 5 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: Synchronously was a first draft at getting it working but I'm going to try a Promise.all and async now 😄

File: data/schema.yaml (hunk 10 above)
Comment (jtart) [reply data/schema.yaml]: Okay. Please remove these changes and add this PR as blocked from merging dependent on https://github.com/bbc/simorgh/pull/380 being merged.

File: src/app/dataValidator/helpers/dataLoader/readScenario.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+const fs = require('fs');
+const path = require('path');
+const { validateData } = require('../validators/validateData');
+
+const readScenario = (fileName, dirName) => {
+  // explicitly ignore scenario-23 as it's a example of invalid data
+  if (fileName.includes('scenario-23.json')) {
```
Comment (jtart) [inline src/app/dataValidator/helpers/dataLoader/readScenario.js]: Rather than explicitly ignoring this specific file, I think it would be better to try and test if an empty object (like scenario-23) and have the ability to ignore if empty object. This is useful if we use this functionality to validate the Optimo data on request.

I think it is fine atm though. Could you maybe put a comment?

**Comment by pjlee11** [comment]
Adding blocked label as this requires #380 to be merged first

File: src/app/dataValidator/helpers/dataLoader/readScenario.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    return false;
+  }
+
+  if (path.extname(fileName) === '.json') {
```
Comment (jtart) [inline src/app/dataValidator/helpers/dataLoader/readScenario.js]: I think we should invert this and `return false` if the ext is not JSON. This would mean the happy path is 
```
return new Promise(resolve => {
      resolve(module.exports.fileToValidate(`${dirName}/${fileName}`));
    });
```

File: src/app/dataValidator/helpers/dataLoader/readScenario.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    return new Promise(resolve => {
+      resolve(module.exports.fileToValidate(`${dirName}/${fileName}`));
```
Comment (jtart) [inline src/app/dataValidator/helpers/dataLoader/readScenario.js]: This doesn't need `module.exports`, and we shouldn't export `fileToValidate`.

File: src/app/dataValidator/index.js [hunk 14]
```diff
@@ -1,16 +1,14 @@
...
+  });
 };
 
-const data = require('../../../data/scenario-01.json');
-
-validateData(data);
+validateAllScenarios();
 
-module.exports.validateData = validateData;
+module.exports = { validateAllScenarios };
```
Comment (jtart) [inline src/app/dataValidator/index.js]: This shouldn't be exported.

File: src/app/dataValidator/helpers/validators/validateData.test.js [hunk 15]
```diff
@@ -0,0 +1,20 @@
+global.console.log = jest.fn(); // silence console.log during jest tests
+global.console.time = jest.fn(); // silence console.time during jest tests
+
+const { validateData } = require('./validateData');
+const data = require('../../../../../data/scenario-01.json');
```
Comment (jtart) [inline src/app/dataValidator/helpers/validators/validateData.test.js]: Should prefix this with `valid` to differentiate between `invalidData`.

File: src/app/dataValidator/helpers/dataLoader/readdirSync.test.js [hunk 16]
```diff
@@ -0,0 +1,12 @@
+global.console.log = jest.fn(); // silence console.log during jest tests
+global.console.time = jest.fn(); // silence console.time during jest tests
+
+const { readdirSync } = require('./readdirSync');
+
+describe('readdirSync helper', () => {
```
Comment (jtart) [inline src/app/dataValidator/helpers/dataLoader/readdirSync.test.js]: Also want to test a promise is returned, and assert that https://github.com/bbc/simorgh/pull/340/files#diff-c0c4410dcc9a30d164183fb60ac9ea02R8 happens.

**Review by jtart — CHANGES_REQUESTED** [review]

File: src/app/dataValidator/helpers/dataLoader/readScenario.js (hunk 13 above)
Comment (pjlee11) [reply src/app/dataValidator/helpers/dataLoader/readScenario.js]: To actively spy on a method in the current module you are testing the method needs to called using the exported version. 

Simple explanation here - https://github.com/facebook/jest/issues/6597

Detailed explanation here - https://github.com/facebook/jest/issues/936#issuecomment-214939935
>The way that jest-mock works is that it runs the module code in isolation and then retrieves the metadata of a module and creates mock functions. Again, in this case it won't have any way to modify the local binding of foo.
>

File: src/app/dataValidator/helpers/dataLoader/readScenario.js (hunk 12 above)
Comment (pjlee11) [reply src/app/dataValidator/helpers/dataLoader/readScenario.js]: 👍

File: src/app/dataValidator/helpers/dataLoader/readScenario.js (hunk 11 above)
Comment (pjlee11) [reply src/app/dataValidator/helpers/dataLoader/readScenario.js]: Happy to open a new issue to address this and merge this PR as is

File: src/app/dataValidator/helpers/validators/validateData.test.js (hunk 15 above)
Comment (pjlee11) [reply src/app/dataValidator/helpers/validators/validateData.test.js]: 👍

File: src/app/dataValidator/index.js (hunk 14 above)
Comment (pjlee11) [reply src/app/dataValidator/index.js]: Is it possible to test this without it exported? If I were just to load the index.js file it wouldn't self execute in the test. I'll have to do some reading into how to properly test self executing files.

File: src/app/dataValidator/helpers/dataLoader/readScenario.js (hunk 11 above)
Comment (pjlee11) [reply src/app/dataValidator/helpers/dataLoader/readScenario.js]: opened issue https://github.com/bbc/simorgh/issues/412

**Comment by pjlee11** [comment]
Output with the `log` messaging silenced.

```
validateBlock: 0.802ms
validateBlock: 0.220ms
validateBlock: 0.353ms
validateBlock: 0.276ms
validateBlock: 0.242ms
validateBlock: 0.241ms
validateBlock: 0.144ms
validateBlock: 0.251ms
validateBlock: 0.278ms
validateBlock: 0.147ms
validateBlock: 0.201ms
validateBlock: 0.380ms
validateBlock: 0.557ms
validateBlock: 0.332ms
validateBlock: 0.408ms
validateBlock: 0.736ms
validateBlock: 2.644ms
validateBlock: 0.587ms
validateBlock: 0.281ms
validateBlock: 0.255ms
validateBlock: 0.150ms
validateBlock: 0.575ms
validateBlock: 0.115ms
validateBlock: 0.902ms
validateBlock: 0.573ms

All 25 scenarios validated!
validateAllScenarios: 20.255ms
```

**Comment by pjlee11** [comment]
Splitting out the data scenario timestamp changes into their own PR - https://github.com/bbc/simorgh/pull/436

**Review by sareh — APPROVED** [review]

**Comment by sareh** [comment]
There are a few changes in terms of naming functions and the output with the silenced `log`, but they are minor points. I think it would be better to merge this in and then make changes at that point, in smaller increments.

**Review by jtart — APPROVED** [review]
